diff --git a/runtime/hidden_api.cc b/runtime/hidden_api.cc
index a20eb664104ed48ae736d622f0a26a6da908da9d..d5830ec3f4b139843eece7df76162c475eebc164 100644
--- a/runtime/hidden_api.cc
+++ b/runtime/hidden_api.cc
@@ -226,6 +226,11 @@ static Domain DetermineDomainFromDexLocation(const std::string& dex_location,
     return Domain::kPlatform;
   }
 
+  if (dex_location == "/system/app/GmsCompatLib/GmsCompatLib.apk" ||
+          (dex_location.starts_with("/data/app/~~") && dex_location.find("==/app.grapheneos.gmscompat.lib-") != std::string::npos)) {
+      return Domain::kPlatform;
+  }
+
   if (class_loader.IsNull()) {
     if (kIsTargetBuild && !kIsTargetLinux) {
       // This is unexpected only when running on Android.
diff --git a/runtime/oat/oat_file_manager.cc b/runtime/oat/oat_file_manager.cc
index 5612c0dcc9f766154c07dfa870c3cb95a5d5cc8a..ef0c296a2469c64df67da93452d5b65f02c55e0b 100644
--- a/runtime/oat/oat_file_manager.cc
+++ b/runtime/oat/oat_file_manager.cc
@@ -488,11 +488,34 @@ std::vector<std::unique_ptr<const DexFile>> OatFileManager::OpenDexFilesFromOat(
   if (dex_files.empty()) {
     std::string error_msg;
     static constexpr bool kVerifyChecksum = true;
-    ArtDexFileLoader dex_file_loader(dex_location);
-    if (!dex_file_loader.Open(Runtime::Current()->IsVerificationEnabled(),
-                              kVerifyChecksum,
-                              /*out*/ &error_msg,
-                              &dex_files)) {
+
+    bool res = false;
+    int fd;
+    if (!strncmp("/gmscompat_fd_", dex_location, strlen("/gmscompat_fd_")) &&
+          sscanf(dex_location, "/gmscompat_fd_%d", &fd) == 1) {
+      fd = dup(fd);
+      struct stat fd_stat;
+      if (fstat(fd, &fd_stat) != 0) {
+          LOG(FATAL) << "fstat of " << dex_location << " failed: " << strerror(errno);
+          _exit(1); // a precaution in case LOG(FATAL) is accidentally or intentionally broken
+      }
+      // Don't allow using /gmscompat_fd to load code from an arbitrary file descriptor, e.g. a
+      // shared memory file descriptor
+      if (!S_ISREG(fd_stat.st_mode)) {
+          LOG(FATAL) << dex_location << " has failed the S_ISREG check, st_mode: 0" << std::oct << fd_stat.st_mode;
+          _exit(1); // a precaution in case LOG(FATAL) is accidentally or intentionally broken
+      }
+      unix_file::FdFile fdFile(fd, false);
+      ArtDexFileLoader dex_file_loader(&fdFile, dex_location);
+      res = dex_file_loader.Open(Runtime::Current()->IsVerificationEnabled(), kVerifyChecksum,
+            /*out*/ &error_msg, &dex_files);
+    } else {
+      ArtDexFileLoader dex_file_loader(dex_location);
+      res = dex_file_loader.Open(Runtime::Current()->IsVerificationEnabled(), kVerifyChecksum,
+            /*out*/ &error_msg, &dex_files);
+    }
+
+    if (!res) {
       ScopedTrace fail_to_open_dex_from_apk("FailedToOpenDexFilesFromApk");
       LOG(WARNING) << error_msg;
       error_msgs->push_back("Failed to open dex files from " + std::string(dex_location)
