diff --git a/api/module-lib-current.txt b/api/module-lib-current.txt
index 5fad4a33232fa2f35f49cdb9ffc871907d4e68a4..6f98aae632e18571871e83effd3111984e0f8edd 100644
--- a/api/module-lib-current.txt
+++ b/api/module-lib-current.txt
@@ -222,6 +222,7 @@ package dalvik.system {
   public final class DelegateLastClassLoader extends dalvik.system.PathClassLoader {
     ctor public DelegateLastClassLoader(String, String, ClassLoader, ClassLoader[]);
     ctor public DelegateLastClassLoader(String, String, ClassLoader, ClassLoader[], ClassLoader[]);
+    field public static java.util.function.BiFunction<java.lang.String,java.lang.Boolean,java.lang.String> modifyClassLoaderPathHook;
   }
 
   public final class DexFile {
@@ -406,6 +407,11 @@ package dalvik.system {
 
 package java.io {
 
+  public class File implements java.lang.Comparable<java.io.File> java.io.Serializable {
+    field public static java.util.function.ToLongFunction<java.io.File> lastModifiedHook;
+    field public static java.util.function.Consumer<java.io.File> mkdirsFailedHook;
+  }
+
   public final class FileDescriptor {
     method public int getInt$();
     method public void setInt$(int);
@@ -557,6 +563,10 @@ package java.text {
 
 package java.util {
 
+  public final class Arrays {
+    field public static java.util.function.UnaryOperator<java.lang.Object[]> asListHook;
+  }
+
   public class LinkedHashMap<K, V> extends java.util.HashMap<K,V> implements java.util.Map<K,V> java.util.SequencedMap<K,V> {
     method public java.util.Map.Entry<K,V> eldest();
   }
@@ -634,6 +644,11 @@ package libcore.io {
     method @NonNull public static java.io.FileDescriptor open(@NonNull String, int) throws java.io.FileNotFoundException;
     method public static int read(@NonNull java.io.FileDescriptor, @NonNull byte[], int, int) throws java.io.IOException;
     method public static void write(@NonNull java.io.FileDescriptor, @NonNull byte[], int, int) throws java.io.IOException;
+    field @Nullable public static libcore.io.IoBridge.OpenFdInterceptor openFdInterceptor;
+  }
+
+  public static interface IoBridge.OpenFdInterceptor {
+    method @Nullable public java.io.FileDescriptor maybeInterceptOpenFd(@NonNull String, int) throws android.system.ErrnoException;
   }
 
   public final class IoUtils {
diff --git a/dalvik/src/main/java/dalvik/system/DelegateLastClassLoader.java b/dalvik/src/main/java/dalvik/system/DelegateLastClassLoader.java
index 2561af71e2069e09d6f121cb42a3dac75f14eaa6..aba982074a0604389dae4f786ed54783d3736090 100644
--- a/dalvik/src/main/java/dalvik/system/DelegateLastClassLoader.java
+++ b/dalvik/src/main/java/dalvik/system/DelegateLastClassLoader.java
@@ -25,6 +25,7 @@ import sun.misc.CompoundEnumeration;
 import java.io.IOException;
 import java.net.URL;
 import java.util.Enumeration;
+import java.util.function.BiFunction;
 
 import libcore.util.NonNull;
 import libcore.util.Nullable;
@@ -43,6 +44,13 @@ import libcore.util.Nullable;
  */
 public final class DelegateLastClassLoader extends PathClassLoader {
 
+    /**
+     * Pre-constructor librarySearchPath hook for GmsCompat
+     * @hide
+     */
+    @SystemApi(client = SystemApi.Client.MODULE_LIBRARIES)
+    public static BiFunction<String, Boolean, String> modifyClassLoaderPathHook;
+
     /**
      * Whether resource loading delegates to the parent class loader. True by default.
      */
@@ -103,7 +111,10 @@ public final class DelegateLastClassLoader extends PathClassLoader {
 
     public DelegateLastClassLoader(@NonNull String dexPath, @Nullable String librarySearchPath,
             @Nullable ClassLoader parent, boolean delegateResourceLoading) {
-        super(dexPath, librarySearchPath, parent);
+        super(
+            maybeModifyClassLoaderPath(dexPath, Boolean.FALSE),
+            maybeModifyClassLoaderPath(librarySearchPath, Boolean.TRUE),
+            parent);
         this.delegateResourceLoading = delegateResourceLoading;
     }
 
@@ -177,6 +188,15 @@ public final class DelegateLastClassLoader extends PathClassLoader {
         this.delegateResourceLoading = true;
     }
 
+    private static String maybeModifyClassLoaderPath(String path, Boolean nativeLibsPath) {
+        BiFunction<String, Boolean, String> hook = modifyClassLoaderPathHook;
+        return hook == null ?
+                path :
+                // replace file paths of GMS Dynamite modules with "/gmscompat_fd_%d" file descriptor
+                // references
+                hook.apply(path, nativeLibsPath);
+    }
+
     @Override
     protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
         // First, check whether the class has already been loaded. Return it if that's the
diff --git a/dalvik/src/main/java/dalvik/system/DexPathList.java b/dalvik/src/main/java/dalvik/system/DexPathList.java
index 780f6b004b3c84faf13844c865dd5696f74eeba3..3b98bc4ec5f0cef37d6d1fc013237993c5a7ef2c 100644
--- a/dalvik/src/main/java/dalvik/system/DexPathList.java
+++ b/dalvik/src/main/java/dalvik/system/DexPathList.java
@@ -377,7 +377,7 @@ public final class DexPathList {
               // We support directories for looking up resources. Looking up resources in
               // directories is useful for running libcore tests.
               elements[elementsPos++] = new Element(file);
-          } else if (file.isFile()) {
+          } else if (file.isFile() || file.getPath().startsWith("/gmscompat_fd_")) {
               String name = file.getName();
 
               DexFile dex = null;
diff --git a/luni/src/main/java/libcore/io/IoBridge.java b/luni/src/main/java/libcore/io/IoBridge.java
index 541e187671d65fd492f15e8f423d82a56d4030b4..6b2df9c323f14a61adee0ec001ba737ed15dd5f1 100644
--- a/luni/src/main/java/libcore/io/IoBridge.java
+++ b/luni/src/main/java/libcore/io/IoBridge.java
@@ -535,6 +535,17 @@ public final class IoBridge {
         }
     }
 
+    /** @hide */
+    @SystemApi(client = MODULE_LIBRARIES)
+    public interface OpenFdInterceptor {
+        @Nullable FileDescriptor maybeInterceptOpenFd(@NonNull String path, int flags) throws ErrnoException;
+    }
+
+    /** @hide */
+    @Nullable
+    @SystemApi(client = MODULE_LIBRARIES)
+    public static OpenFdInterceptor openFdInterceptor;
+
     /**
      * Wrapper for {@link Os#open(String, int, int)} that behaves similar to {@link java.io.File}.
      * When a {@link java.io.File} is opened and there is an error, it throws
@@ -557,7 +568,13 @@ public final class IoBridge {
     public static @NonNull FileDescriptor open(@NonNull String path, int flags) throws FileNotFoundException {
         FileDescriptor fd = null;
         try {
-            fd = Libcore.os.open(path, flags, 0666);
+            OpenFdInterceptor interceptor = openFdInterceptor;
+            if (interceptor != null) {
+                fd = interceptor.maybeInterceptOpenFd(path, flags);
+            }
+            if (fd == null) {
+                fd = Libcore.os.open(path, flags, 0666);
+            }
             // Posix open(2) fails with EISDIR only if you ask for write permission.
             // Java disallows reading directories too.f
             if (S_ISDIR(Libcore.os.fstat(fd).st_mode)) {
diff --git a/luni/src/main/native/libcore_io_Linux.cpp b/luni/src/main/native/libcore_io_Linux.cpp
index 78f1834cd52530a1f7e0c5e76e73c9e16011906c..944b8bf7b7d4f39f078fb0ba03b89c7e80920464 100644
--- a/luni/src/main/native/libcore_io_Linux.cpp
+++ b/luni/src/main/native/libcore_io_Linux.cpp
@@ -803,8 +803,17 @@ static jobject doStat(JNIEnv* env, jstring javaPath, bool isLstat) {
         return NULL;
     }
     struct stat64 sb;
-    int rc = isLstat ? TEMP_FAILURE_RETRY(lstat64(path.c_str(), &sb))
+
+    int rc;
+    int gmscompat_fd;
+    if (strncmp("/gmscompat_fd_", path.c_str(), strlen("/gmscompat_fd_")) == 0
+            && sscanf(path.c_str(), "/gmscompat_fd_%d", &gmscompat_fd) == 1) {
+        rc = TEMP_FAILURE_RETRY(fstat64(gmscompat_fd, &sb));
+    } else {
+        rc = isLstat ? TEMP_FAILURE_RETRY(lstat64(path.c_str(), &sb))
                      : TEMP_FAILURE_RETRY(stat64(path.c_str(), &sb));
+    }
+
     if (rc == -1) {
         throwErrnoException(env, isLstat ? "lstat" : "stat");
         return NULL;
@@ -1991,7 +2000,15 @@ static jobject Linux_open(JNIEnv* env, jobject, jstring javaPath, jint flags, ji
     if (path.c_str() == NULL) {
         return NULL;
     }
-    int fd = throwIfMinusOne(env, "open", TEMP_FAILURE_RETRY(open(path.c_str(), flags, mode)));
+    int ret;
+    int gmscompat_fd;
+    if (strncmp("/gmscompat_fd_", path.c_str(), strlen("/gmscompat_fd_")) == 0
+            && sscanf(path.c_str(), "/gmscompat_fd_%d", &gmscompat_fd) == 1) {
+        ret = dup(gmscompat_fd);
+    } else {
+        ret = open(path.c_str(), flags, mode);
+    }
+    int fd = throwIfMinusOne(env, "open", ret);
     return createFileDescriptorIfOpen(env, fd);
 }
 
diff --git a/ojluni/src/main/java/java/io/File.java b/ojluni/src/main/java/java/io/File.java
index 3b921a34fdc0002ad1591711465bde731211c527..8bba36ab62e7eb1e742b840b3a0cca7b559a6b6c 100644
--- a/ojluni/src/main/java/java/io/File.java
+++ b/ojluni/src/main/java/java/io/File.java
@@ -26,6 +26,7 @@
 
 package java.io;
 
+import android.annotation.SystemApi;
 import android.compat.annotation.ChangeId;
 import android.compat.annotation.EnabledSince;
 
@@ -40,6 +41,9 @@ import java.util.ArrayList;
 import java.security.AccessController;
 import java.nio.file.Path;
 import java.nio.file.FileSystems;
+import java.util.function.Consumer;
+import java.util.function.ToLongFunction;
+
 import sun.security.action.GetPropertyAction;
 
 // Android-added: Info about UTF-8 usage in filenames.
@@ -166,6 +170,20 @@ public class File
      */
     private static final FileSystem fs = DefaultFileSystem.getFileSystem();
 
+    /**
+     * File#lastModified() hook for GmsCompat
+     * @hide
+     */
+    @SystemApi(client = SystemApi.Client.MODULE_LIBRARIES)
+    public static ToLongFunction<File> lastModifiedHook;
+
+    /**
+     * File#mkdirs() hook for GmsCompat
+     * @hide
+     */
+    @SystemApi(client = SystemApi.Client.MODULE_LIBRARIES)
+    public static Consumer<File> mkdirsFailedHook;
+
     /**
      * This abstract pathname's normalized pathname string. A normalized
      * pathname string uses the default name-separator character and does not
@@ -950,7 +968,14 @@ public class File
         if (isInvalid()) {
             return 0L;
         }
-        return fs.getLastModifiedTime(this);
+        long res = fs.getLastModifiedTime(this);
+        if (res == 0L) {
+            ToLongFunction<File> hook = lastModifiedHook;
+            if (hook != null) {
+                res = hook.applyAsLong(this);
+            }
+        }
+        return res;
     }
 
     /**
@@ -1374,8 +1399,16 @@ public class File
         }
 
         File parent = canonFile.getParentFile();
-        return (parent != null && (parent.mkdirs() || parent.exists()) &&
+        boolean res = (parent != null && (parent.mkdirs() || parent.exists()) &&
                 canonFile.mkdir());
+
+        if (!res) {
+            Consumer<File> hook = mkdirsFailedHook;
+            if (hook != null) {
+                hook.accept(this);
+            }
+        }
+        return res;
     }
 
     // Android-changed: Replaced generic platform info with Android specific one.
diff --git a/ojluni/src/main/java/java/util/Arrays.java b/ojluni/src/main/java/java/util/Arrays.java
index 3ed26f73d0b1ec3a9df681ee66a29352f5a89140..3e5b066f0987c28b5928a0952576dff229690ee5 100644
--- a/ojluni/src/main/java/java/util/Arrays.java
+++ b/ojluni/src/main/java/java/util/Arrays.java
@@ -4521,9 +4521,17 @@ public final class Arrays {
     @SafeVarargs
     @SuppressWarnings("varargs")
     public static <T> List<T> asList(T... a) {
+        java.util.function.UnaryOperator<Object[]> hook = asListHook;
+        if (hook != null) {
+            a = (T[]) hook.apply(a);
+        }
         return new ArrayList<>(a);
     }
 
+    /** @hide */
+    @android.annotation.SystemApi(client = android.annotation.SystemApi.Client.MODULE_LIBRARIES)
+    public static java.util.function.UnaryOperator<Object[]> asListHook;
+
     /**
      * Since Android 15 Arrays.asList(...).toArray()'s component type is {@link Object},
      * not the underlying array's elements type. So the following code will throw
diff --git a/ojluni/src/main/java/java/util/zip/ZipFile.java b/ojluni/src/main/java/java/util/zip/ZipFile.java
index 9ab6713c9e8f7f0c2319d614eb04eda0d9d2e668..1c319c55beff54b0e2294d25bfbccac0edabd46c 100644
--- a/ojluni/src/main/java/java/util/zip/ZipFile.java
+++ b/ojluni/src/main/java/java/util/zip/ZipFile.java
@@ -1523,6 +1523,7 @@ public class ZipFile implements ZipConstants, Closeable {
             public Key(File file, BasicFileAttributes attrs, ZipCoder zc,
                     boolean isZipFilePathValidatorEnabled) {
                 this.attrs = attrs;
+
                 this.file = file;
                 this.utf8 = zc.isUTF8();
                 this.isZipFilePathValidatorEnabled = isZipFilePathValidatorEnabled;
@@ -1640,6 +1641,7 @@ public class ZipFile implements ZipConstants, Closeable {
             } else {
                 // Android-changed: open with O_CLOEXEC flag set.
                 // this.zfile = new RandomAccessFile(key.file, "r");
+                // todo
                 this.zfile = new RandomAccessFile(key.file, "r", /* setCloExecFlag= */ true);
             }
             try {
diff --git a/ojluni/src/main/native/UnixFileSystem_md.c b/ojluni/src/main/native/UnixFileSystem_md.c
index 02dd29794bc862f68c620b05515e3abe085ba211..658e68d05b3b0834f6eb594bb6908dd627b34ff9 100644
--- a/ojluni/src/main/native/UnixFileSystem_md.c
+++ b/ojluni/src/main/native/UnixFileSystem_md.c
@@ -252,7 +252,16 @@ Java_java_io_UnixFileSystem_getLastModifiedTime0(JNIEnv *env, jobject this,
 
     WITH_FIELD_PLATFORM_STRING(env, file, ids.path, path) {
         struct stat64 sb;
-        if (stat64(path, &sb) == 0) {
+        int ret;
+        int gmscompat_fd;
+        if (strncmp("/gmscompat_fd_", path, strlen("/gmscompat_fd_")) == 0
+                && sscanf(path, "/gmscompat_fd_%d", &gmscompat_fd) == 1) {
+            ret = fstat64(gmscompat_fd, &sb);
+        } else {
+            ret = stat64(path, &sb);
+        }
+
+        if (ret == 0) {
 #if defined(_AIX)
             rv =  (jlong)sb.st_mtime * 1000;
             rv += (jlong)sb.st_mtime_n / 1000000;
diff --git a/ojluni/src/main/native/UnixNativeDispatcher.c b/ojluni/src/main/native/UnixNativeDispatcher.c
index 2de1ad428cca8bed9edeab9f62b37f6e1edb4966..1d1cd512b5843fb9d9de792e11dff372170c1aef 100644
--- a/ojluni/src/main/native/UnixNativeDispatcher.c
+++ b/ojluni/src/main/native/UnixNativeDispatcher.c
@@ -23,6 +23,7 @@
  * questions.
  */
 
+#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <limits.h>
@@ -542,6 +543,17 @@ static void prepAttributes(JNIEnv* env, struct stat64* buf, jobject attrs) {
 #endif
 }
 
+static bool interceptGmsCompatFdStat(JNIEnv* env, jclass this, const char* path, jobject attrs) {
+    int fd;
+    if (strncmp("/gmscompat_fd_", path, strlen("/gmscompat_fd_")) == 0
+            && sscanf(path, "/gmscompat_fd_%d", &fd) == 1) {
+        Java_sun_nio_fs_UnixNativeDispatcher_fstat(env, this, fd, attrs);
+        return true;
+    }
+
+    return false;
+}
+
 JNIEXPORT void JNICALL
 Java_sun_nio_fs_UnixNativeDispatcher_stat0(JNIEnv* env, jclass this,
     jlong pathAddress, jobject attrs)
@@ -550,6 +562,10 @@ Java_sun_nio_fs_UnixNativeDispatcher_stat0(JNIEnv* env, jclass this,
     struct stat64 buf;
     const char* path = (const char*)jlong_to_ptr(pathAddress);
 
+    if (interceptGmsCompatFdStat(env, this, path, attrs)) {
+        return;
+    }
+
     RESTARTABLE(stat64(path, &buf), err);
     if (err == -1) {
         throwUnixException(env, errno);
@@ -580,6 +596,10 @@ Java_sun_nio_fs_UnixNativeDispatcher_lstat0(JNIEnv* env, jclass this,
     struct stat64 buf;
     const char* path = (const char*)jlong_to_ptr(pathAddress);
 
+    if (interceptGmsCompatFdStat(env, this, path, attrs)) {
+        return;
+    }
+
     RESTARTABLE(lstat64(path, &buf), err);
     if (err == -1) {
         throwUnixException(env, errno);
