diff --git a/framework-t/src/android/app/usage/NetworkStatsManager.java b/framework-t/src/android/app/usage/NetworkStatsManager.java
index 18c839f68e742ed3ab93c75ed2889e02fb51c735..eeb9806ba35c2ceee535c33a9e7cf4df8327455b 100644
--- a/framework-t/src/android/app/usage/NetworkStatsManager.java
+++ b/framework-t/src/android/app/usage/NetworkStatsManager.java
@@ -31,6 +31,7 @@ import android.annotation.RequiresPermission;
 import android.annotation.SystemApi;
 import android.annotation.SystemService;
 import android.annotation.WorkerThread;
+import android.app.compat.gms.GmsCompat;
 import android.app.usage.NetworkStats.Bucket;
 import android.compat.annotation.UnsupportedAppUsage;
 import android.content.Context;
@@ -593,8 +594,19 @@ public class NetworkStatsManager {
     @WorkerThread
     public NetworkStats queryDetailsForUidTag(int networkType, @Nullable String subscriberId,
             long startTime, long endTime, int uid, int tag) throws SecurityException {
-        return queryDetailsForUidTagState(networkType, subscriberId, startTime, endTime, uid,
-            tag, NetworkStats.Bucket.STATE_ALL);
+        try {
+            return queryDetailsForUidTagState(networkType, subscriberId, startTime, endTime, uid,
+                tag, NetworkStats.Bucket.STATE_ALL);
+        } catch (SecurityException e) {
+            if (GmsCompat.isEnabled()) {
+                // GmsCore null-checks the return value despite the method being marked as non-null.
+                // An alternative to returning null is to make a stub instance of NetworkStats, but
+                // it's complex because of the way NetworkStats is implemented.
+                //noinspection DataFlowIssue
+                return null;
+            }
+            throw e;
+        }
     }
 
     /**
diff --git a/framework/src/android/net/ConnectivityManager.java b/framework/src/android/net/ConnectivityManager.java
index 1a07d33f0740bdba8b86a68f8f7ee3a0cb06b8ee..35df72a413620c3ea4333d985b79742f1be56f90 100644
--- a/framework/src/android/net/ConnectivityManager.java
+++ b/framework/src/android/net/ConnectivityManager.java
@@ -41,6 +41,7 @@ import android.annotation.SystemApi;
 import android.annotation.SystemService;
 import android.app.PendingIntent;
 import android.app.admin.DevicePolicyManager;
+import android.app.compat.gms.GmsCompat;
 import android.compat.annotation.UnsupportedAppUsage;
 import android.content.ComponentName;
 import android.content.Context;
@@ -3151,6 +3152,10 @@ public class ConnectivityManager {
     @RequiresPermission(anyOf = {android.Manifest.permission.TETHER_PRIVILEGED,
             android.Manifest.permission.WRITE_SETTINGS})
     public boolean isTetheringSupported() {
+        if (GmsCompat.isEnabled()) {
+            return false;
+        }
+
         return getTetheringManager().isTetheringSupported();
     }
 
diff --git a/nearby/framework/java/android/nearby/NearbyManager.java b/nearby/framework/java/android/nearby/NearbyManager.java
index 9d7d144a385ab42276ba8c8ef6e1a755b68e27d0..f64a684d05aa95c1fd0316227d2232181b8d75ae 100644
--- a/nearby/framework/java/android/nearby/NearbyManager.java
+++ b/nearby/framework/java/android/nearby/NearbyManager.java
@@ -26,6 +26,7 @@ import android.annotation.RequiresPermission;
 import android.annotation.SuppressLint;
 import android.annotation.SystemApi;
 import android.annotation.SystemService;
+import android.app.compat.gms.GmsCompat;
 import android.bluetooth.BluetoothManager;
 import android.content.Context;
 import android.location.LocationManager;
@@ -220,6 +221,9 @@ public class NearbyManager {
             }
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
+        } catch (SecurityException e) {
+            GmsCompat.catchOrRethrow(e);
+            return ScanStatus.ERROR;
         }
     }
 
@@ -255,6 +259,8 @@ public class NearbyManager {
             }
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
+        } catch (SecurityException e) {
+            GmsCompat.catchOrRethrow(e);
         }
     }
 
@@ -509,6 +515,8 @@ public class NearbyManager {
             mService.setPoweredOffFindingEphemeralIds(ephemeralIdList);
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
+        } catch (SecurityException e) {
+            GmsCompat.catchOrRethrow(e);
         }
 
     }
@@ -554,6 +562,8 @@ public class NearbyManager {
                     poweredOffFindingMode == POWERED_OFF_FINDING_MODE_ENABLED);
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
+        } catch (SecurityException e) {
+            GmsCompat.catchOrRethrow(e);
         }
     }
 
@@ -577,6 +587,9 @@ public class NearbyManager {
                     ? POWERED_OFF_FINDING_MODE_ENABLED : POWERED_OFF_FINDING_MODE_DISABLED;
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
+        } catch (SecurityException e) {
+            GmsCompat.catchOrRethrow(e);
+            return POWERED_OFF_FINDING_MODE_DISABLED;
         }
     }
 
