diff --git a/framework/java/android/nfc/NfcAdapter.java b/framework/java/android/nfc/NfcAdapter.java
index 28ff82602219a776e45c8526788538f8ac15a381..1cb995bbf045643ef09b77138bc6b0f2b14b1a6a 100644
--- a/framework/java/android/nfc/NfcAdapter.java
+++ b/framework/java/android/nfc/NfcAdapter.java
@@ -32,6 +32,8 @@ import android.annotation.TestApi;
 import android.annotation.UserIdInt;
 import android.app.Activity;
 import android.app.PendingIntent;
+import android.app.compat.gms.GmsCompat;
+import android.app.compat.gms.GmsModuleHooks;
 import android.compat.annotation.UnsupportedAppUsage;
 import android.content.Context;
 import android.content.IntentFilter;
@@ -1141,6 +1143,13 @@ public final class NfcAdapter {
     @FlaggedApi(Flags.FLAG_NFC_STATE_CHANGE)
     @RequiresPermission(android.Manifest.permission.WRITE_SECURE_SETTINGS)
     public boolean enable() {
+        if (GmsCompat.isEnabled()) {
+            if (!GmsCompat.hasPermission(android.Manifest.permission.WRITE_SECURE_SETTINGS)) {
+                GmsModuleHooks.enableNfc();
+                return false;
+            }
+        }
+
         return callServiceReturn(() ->  sService.enable(mContext.getPackageName()), false);
 
     }
