diff --git a/framework/java/android/app/StatsManager.java b/framework/java/android/app/StatsManager.java
index ddee2c423866ff539890cd446fc6193d9e90263f..76ddb8796691ccb2a975981da099de7c1a1e1364 100644
--- a/framework/java/android/app/StatsManager.java
+++ b/framework/java/android/app/StatsManager.java
@@ -25,6 +25,7 @@ import android.annotation.NonNull;
 import android.annotation.Nullable;
 import android.annotation.RequiresPermission;
 import android.annotation.SystemApi;
+import android.app.compat.gms.GmsCompat;
 import android.content.Context;
 import android.os.Binder;
 import android.os.Build;
@@ -168,6 +169,9 @@ public final class StatsManager {
      */
     @RequiresPermission(allOf = { DUMP, PACKAGE_USAGE_STATS })
     public void addConfig(long configKey, byte[] config) throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -214,6 +218,9 @@ public final class StatsManager {
      */
     @RequiresPermission(allOf = { DUMP, PACKAGE_USAGE_STATS })
     public void removeConfig(long configKey) throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -283,6 +290,9 @@ public final class StatsManager {
     public void setBroadcastSubscriber(
             PendingIntent pendingIntent, long configKey, long subscriberId)
             throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -337,6 +347,9 @@ public final class StatsManager {
     @RequiresPermission(allOf = { DUMP, PACKAGE_USAGE_STATS })
     public void setFetchReportsOperation(PendingIntent pendingIntent, long configKey)
             throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -374,6 +387,9 @@ public final class StatsManager {
     @RequiresPermission(allOf = { DUMP, PACKAGE_USAGE_STATS })
     public @NonNull long[] setActiveConfigsChangedOperation(@Nullable PendingIntent pendingIntent)
             throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return new long[0];
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -422,6 +438,9 @@ public final class StatsManager {
             @NonNull String configPackage,
             @Nullable PendingIntent pendingIntent)
             throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return new long[0];
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -482,6 +501,9 @@ public final class StatsManager {
             @NonNull @CallbackExecutor Executor executor,
             @NonNull OutcomeReceiver<StatsCursor, StatsQueryException> outcomeReceiver)
             throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         if(query.getSqlDialect() != StatsQuery.DIALECT_SQLITE) {
             executor.execute(() -> {
                 outcomeReceiver.onError(new StatsQueryException("Unsupported Sql Dialect"));
@@ -530,6 +552,9 @@ public final class StatsManager {
      */
     @RequiresPermission(allOf = { DUMP, PACKAGE_USAGE_STATS })
     public byte[] getReports(long configKey) throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return new byte[0];
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -575,6 +600,9 @@ public final class StatsManager {
      */
     @RequiresPermission(allOf = { DUMP, PACKAGE_USAGE_STATS })
     public byte[] getStatsMetadata() throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return new byte[0];
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -615,6 +643,9 @@ public final class StatsManager {
     @RequiresPermission(allOf = {DUMP, PACKAGE_USAGE_STATS})
     public long[] getRegisteredExperimentIds()
             throws StatsUnavailableException {
+        if (GmsCompat.isEnabled()) {
+            return new long[0];
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
@@ -651,6 +682,9 @@ public final class StatsManager {
     public void setPullAtomCallback(int atomTag, @Nullable PullAtomMetadata metadata,
             @NonNull @CallbackExecutor Executor executor,
             @NonNull StatsPullAtomCallback callback) {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         long coolDownMillis =
                 metadata == null ? DEFAULT_COOL_DOWN_MILLIS : metadata.mCoolDownMillis;
         long timeoutMillis = metadata == null ? DEFAULT_TIMEOUT_MILLIS : metadata.mTimeoutMillis;
@@ -681,6 +715,9 @@ public final class StatsManager {
      */
     @RequiresPermission(android.Manifest.permission.REGISTER_STATS_PULL_ATOM)
     public void clearPullAtomCallback(int atomTag) {
+        if (GmsCompat.isEnabled()) {
+            return;
+        }
         synchronized (sLock) {
             try {
                 IStatsManagerService service = getIStatsManagerServiceLocked();
