diff --git a/src/com/android/providers/downloads/DownloadProvider.java b/src/com/android/providers/downloads/DownloadProvider.java
index f3f3bc3182280e45184e8e89bb0b571dedaa112f..a06adc9827ee382cf7e14c8be3cd8ae172066771 100644
--- a/src/com/android/providers/downloads/DownloadProvider.java
+++ b/src/com/android/providers/downloads/DownloadProvider.java
@@ -1223,6 +1223,14 @@ public final class DownloadProvider extends ContentProvider {
         values.remove(Downloads.Impl.COLUMN_MIME_TYPE);
         values.remove(Downloads.Impl.COLUMN_FILE_NAME_HINT); // checked later in insert()
         values.remove(Downloads.Impl.COLUMN_NOTIFICATION_PACKAGE); // checked later in insert()
+        // GmsCompat: Play Store uses COLUMN_NOTIFICATION_CLASS as selectionArg
+        // in ContentResolver#query() during self-update.
+        // Allowing this column is harmless
+        // (used only for Intent.setClassName(packageName, className)) and
+        // it doesn't even get used anyway, because DownloadProvider always checks whether
+        // COLUMN_IS_PUBLIC_API is "false" (it is enforced to be "true" above) before
+        // using COLUMN_NOTIFICATION_CLASS
+        values.remove(Downloads.Impl.COLUMN_NOTIFICATION_CLASS);
         values.remove(Downloads.Impl.COLUMN_ALLOWED_NETWORK_TYPES);
         values.remove(Downloads.Impl.COLUMN_ALLOW_ROAMING);
         values.remove(Downloads.Impl.COLUMN_ALLOW_METERED);
