diff --git a/generate_java_binder.cpp b/generate_java_binder.cpp
index 5b37d2191b881e82ef84982982cb99fe6ab413bf..29e5198b3f398e88715a41326586db08746b0f62 100644
--- a/generate_java_binder.cpp
+++ b/generate_java_binder.cpp
@@ -387,19 +387,21 @@ class ProxyClass : public Class {
 };
 
 ProxyClass::ProxyClass(const AidlInterface* interfaceType, const Options& options) : Class() {
-  this->modifiers = PRIVATE | STATIC;
+  this->modifiers = PUBLIC | STATIC;
+  this->comment = "/** @hide */";
   this->what = Class::CLASS;
   this->type = interfaceType->GetCanonicalName() + ".Stub.Proxy";
   this->interfaces.push_back(interfaceType->GetCanonicalName());
 
   // IBinder mRemote
   mRemote = std::make_shared<Variable>("android.os.IBinder", "mRemote");
-  this->elements.push_back(std::make_shared<Field>(PRIVATE, mRemote));
+  this->elements.push_back(std::make_shared<Field>(PROTECTED, mRemote));
 
   // Proxy()
   auto remote = std::make_shared<Variable>("android.os.IBinder", "remote");
   auto ctor = std::make_shared<Method>();
   ctor->name = "Proxy";
+  ctor->modifiers = PROTECTED;
   ctor->statements = std::make_shared<StatementBlock>();
   ctor->parameters.push_back(remote);
   ctor->statements->Add(std::make_shared<Assignment>(mRemote, remote));
@@ -919,8 +921,9 @@ static void GenerateMethods(const AidlInterface& iface, const AidlMethod& method
   string transactCodeName = "TRANSACTION_";
   transactCodeName += method.GetName();
 
-  auto transactCode =
-      std::make_shared<Field>(STATIC | FINAL, std::make_shared<Variable>("int", transactCodeName));
+  auto transactCode = std::make_shared<Field>(
+      PUBLIC | STATIC | FINAL, std::make_shared<Variable>("int", transactCodeName));
+  transactCode->comment = "/** @hide */";
   transactCode->value =
       StringPrintf("(android.os.IBinder.FIRST_CALL_TRANSACTION + %d)", index);
   stubClass->elements.push_back(transactCode);
